### DISCUSSION TITLE
Cool tips you want to share

### DISCUSSION THREAD

**Discussion by ekickx** [ekickx/clipboard-image.nvim#15] [Show and tell]
Sharing is nice, you can show some of your trick/hack of this plugin here

**Comment by ekickx** [comment]
### Manually insert image name

<details>
<summary><strong>Demo</strong></summary>

![demo](https://user-images.githubusercontent.com/26477782/129376906-a29ea344-dbc0-483f-8b70-55b5476e85e5.gif)
</details>

<details open>
<summary><strong>Config</strong></summary>

```lua
require'clipboard-image'.setup {
  default = {
    img_name = function ()
      vim.fn.inputsave()
      local name = vim.fn.input('Name: ')
      vim.fn.inputrestore()
      return name
    end,
  }
}
```
</details>

**Reply by erwin** [reply]
If `fn.input` is empty, might be nice to set `img_name` to your `os.date('%y-%m-%d-%H-%M-%S')` default value

**Reply by ekickx** [reply]
That's a good idea. Then you can do this

```lua
require'clipboard-image'.setup {
  default = {
    img_name = function ()
      vim.fn.inputsave()
      local name = vim.fn.input('Name: ')
      vim.fn.inputrestore()

      if name == nil or name == '' then
        return os.date('%y-%m-%d-%H-%M-%S')
      end
      return name
    end,
  }
}

```

**Reply by erwin** [reply]
Any idea how to get this working with `vim.ui.input` instead of `vim.fn.input`?

The `vim.ui.input` can look like this: (may require `stevearc/dressing.nvim`)
![image](https://user-images.githubusercontent.com/11534217/204448639-d311492e-0619-4c12-b35a-b16604818318.png)

I found this post saying `vim.ui.input` can be a drop in replacement...
https://github.com/nvim-telescope/telescope.nvim/issues/1479#issue-1059815789

Yet I've tried about a dozen permutations and haven't been able to get it to work...

```
local async = require("plenary.async")
-- local input = vim.fn.input
local input = async.wrap(function(prompt, text, completion, callback)
    vim.ui.input({
        prompt = prompt,
        default = text,
        completion = completion,
    }, callback)
end, 4)
```

**Reply by ekickx** [reply]
This might work:

```lua
require("clipboard-image").setup {
  markdown = {
    img_name = function ()
      local name = os.date('%y-%m-%d-%H-%M-%S')

      vim.ui.input({prompt = "Name: ", default = name}, function(input)
        name = input
      end)
      return name
    end
  }
}
```

**Reply by erwin** [reply]
Thank you very much for trying.
Unfortunately that didn't work for me - it appears to have fired the prompt after the file was written.

I guess I need to spend some more time with async in Lua.

**Reply by ekickx** [reply]
EH really? Can you tell me your neovim version. Because it works on my neovim (0.7)

**Reply by ekickx** [reply]
Or do you change other config option? Maybe it affects the result(?)

**Reply by erwin** [reply]
I'm on nvim v0.8.1... Great to hear it works for you though... Maybe some other plugin is conflicting with it.

**Reply by erwin** [reply]
@ekickx you are correct... by disabling the `dressing.nvim` plugin, it does work exactly as you wrote.

https://github.com/stevearc/dressing.nvim

Unfortunately, the reason I wanted to use `vim.ui.input` was for these pop-up dialogs... Back to the drawing board :-)

![image](https://user-images.githubusercontent.com/11534217/204734355-98e0221d-bd41-42ec-b56e-28f4d008f4fb.png)

**Reply by kris-swann** [reply]
This was exactly what I was looking for, thanks so much! Using this as a base I was able to add support for using selected text or prompting for a name depending on the mode

```lua
require'clipboard-image'.setup {
  default = {
    img_name = function ()
      -- If in visual mode use selected text
      local mode = vim.api.nvim_get_mode().mode
      if mode == 'v' then
        -- Recover visual selection by: `v` to exit visual mode then `gv`
        -- (If you don't exit then gv will select your previous selection)
        -- finally yank the the `a` register
        vim.cmd("normal! vgv\"ay")
        -- Evaluate contents of `a` register
        local selection = vim.api.nvim_eval("@a")
        -- Delete the selection (will be replaced by clipboard-image)
        vim.cmd("normal! gvd")
        return selection
      else
        -- Prompt for name
        vim.fn.inputsave()
        local name = vim.fn.input('Image Name: ')
        vim.fn.inputrestore()
        return name
      end
    end,
  },
}

vim.keymap.set({ 'n', 'v' }, '<leader><C-p>', '<cmd>PasteImg<CR>')

```

Credit to https://github.com/neovim/neovim-ruby/issues/66#issuecomment-633711688 for the visual register trick

**Reply by windowsoffender** [reply]
@erwin Did you manage to get it working with `vim.ui.input` and `dressing.nvim`?

**Comment by jeffizz** [comment]
```lua
lua <<LUA
--
require'clipboard-image'.setup {
  -- Default configuration for all typefile
  default = {
    img_dir = "img",
    img_dir_txt = "img",
    img_name = function () return os.date('%Y-%m-%d-%H-%M-%S') end,
    affix = "![](/%s)",
    img_handler = function (img)
        local script = string.format('~/bin/imgup %s &', img.path)
        os.execute(script)
    end
  },
  -- You can create configuration for ceartain filetype by creating another field (markdown, in this case)
  -- If you're uncertain what to name your field to, you can run `:set filetype?`
  -- Missing options from `markdown` field will be replaced by options from `default` field
  markdown = {
    img_dir = "img",
    img_dir_txt = "img",
    affix = "![](/%s)"
  }
}

LUA
command! -nargs=* PasteImage call luaeval('
    \ require"clipboard-image.paste".paste_img({
        \ img_dir = "~/Assets/"..(_A[1] and _A[1] or os.date("/%Y/%m")),
        \ img_dir_txt = (_A[1] and _A[1] or os.date("/%Y/%m")),
        \ img_name = os.date("%d%H%M%S")..(_A[2] or "")
    \ })
    \ ', split('<args>'))

noremap <Leader>P :<C-U><C-R>="PasteImage "<CR>
```

**Reply by njnygaard** [reply]
`imgup` is a great hint, thanks!

**Comment by ekickx** [comment]
### Image directory relative to current file

<details>
<summary><strong>Demo</strong></summary>

https://user-images.githubusercontent.com/26477782/148644100-8cb06bde-0de0-4d58-bf37-7c0ae955ff91.mp4
</details>

<details open>
<summary><strong>Config</strong></summary>

```lua
require'clipboard-image'.setup {
  default = {
   img_dir = {"%:p:h", "img"}
  }
}
```
</details>

`%:p:h` will get the directory of your current file. See also `:help cmdline-special` and `:help filename-modifiers`

**Reply by larntz** [reply]
this should be the default imo

**Reply by danielwe** [reply]
Building on this, if you want to place the image in a subfolder under `img` named after the current file:
```lua
require'clipboard-image'.setup {
  default = {
   img_dir = {"%:p:h", "img", "%:t:r"}
  }
}

**Reply by danielwe** [reply]
One issue with this is if the expanded path contains spaces or other special characters. To form valid markdown or html links you need to percent encode these (e.g., a space should be `%20`). You can take care of this using a small plugin I wrote, https://github.com/danielwe/vim-percent, with the following config:

```lua
require'clipboard-image'.setup {
  default = {
    img_dir = {"%:p:h", "img"},
    img_dir_txt = function()
      return {vim.fn["percent#encode"](vim.fn.expand("%:p:h")), "img"}
    end,
    affix = "![](%s)",
  }
}
```

**Reply by tan-wei** [reply]
Fabulous!

**Comment by ekickx** [comment]
### Insert alt-text with `img_handler`

<details>
<summary><strong>Demo</strong></summary>


https://user-images.githubusercontent.com/26477782/153805168-f794d46f-2bdf-4091-827e-a7cc02e4acf4.mp4


</details>

<details open>
<summary><strong>Config</strong></summary>

```lua
require 'clipboard-image'.setup {
  markdown = {
    img_handler = function(img)
      vim.cmd("normal! f[") -- go to [
      vim.cmd("normal! a" .. img.name) -- append text with image name
    end
  }
}
```
</details>

`:normal` executes key mapping in normal mode. See also `:h :normal`

**Comment by ekickx** [comment]
### Use word under cursor as image's name
This tip by @gokulmaxi from #14 

<details>
<summary><strong>Demo</strong></summary>



![](https://raw.githubusercontent.com/ekickx/clipboard-image.nvim/c86543c8b0bb9917e63adb66aa8a5147c735b431/img/wordUnderCursor.gif)


</details>

</details>

<details open>
<summary><strong>Config</strong></summary>

```lua
require'clipboard-image'.setup {
  tex = {
    img_name = function()
      local cword = vim.fn.expand('<cword>')
      vim.fn.feedkeys("diw")
      return cword      
    end,
  }
}
```
</details>

**Comment by ekickx** [comment]
### Copy the URL instead of the actual image

<details>
<summary><strong>Demo</strong></summary>

https://user-images.githubusercontent.com/26477782/167260067-13604f9e-415d-4257-91f5-d223c289fa1c.mp4

</details>

</details>

<details open>
<summary><strong>Config</strong></summary>

```lua
local function paste_url(url)
  url = url.args
  local utils = require "clipboard-image.utils"
  local conf_utils = require "clipboard-image.config"

  local conf_toload = conf_utils.get_usable_config()
  local conf = conf_utils.load_config(conf_toload)

  utils.insert_txt(conf.affix, url)
end

-- Now let's create the command (works on neovim 0.7+)
local create_command = vim.api.nvim_create_user_command
create_command("PasteImgUrl", paste_url, { nargs=1 })
```
</details>

**Comment by erwin** [comment]
Unfortunately I did not.
May have been possible, but I ran out of time to spend on it.
